france_connect/clients.py

```diff
@@ -0,0 +1,169 @@
+        fc_authorization_url: str = "/api/v2/authorize",
+        fc_token_url: str = "/api/v2/token",
+        fc_jwks_url: str = "/api/v2/jwks",
+        fc_userinfo_url: str = "/api/v2/userinfo",
+        fc_logout_url: str = "/api/v2/session/end",
+    ):
+        self.client_id = client_id
+        self.client_secret = client_secret
+        self.scopes = scopes
+        self.post_login_callback_url = post_login_callback_url
+        self.post_logout_callback_url = post_logout_callback_url
+        self.fc_base_url = fc_base_url
+        self.fc_authorization_url = fc_authorization_url
+        self.fc_token_url = fc_token_url
+        self.fc_jwks_url = fc_jwks_url
+        self.fc_userinfo_url = fc_userinfo_url
+        self.fc_logout_url = fc_logout_url
+
+    def generate_nonce(self) -> str:
+        """Generate a random nonce in hexadecimal string (64 characters).."""
+        return secrets.token_hex(48)
+
+    def generate_state(self) -> str:
+        """Generate a random state in hexadecimal string (64 characters)."""
+        return secrets.token_hex(48)
+
+    def get_authentication_url(
+        self, acr_values: List[ACRValues] = [ACRValues.EIDAS1.value], nonce: str = None, state: str = None
+    ) -> Tuple[str, str, str]:
+        """Return the authentication URL, nonce and state."""
+        nonce = nonce or self.generate_nonce()
+        state = state or self.generate_state()
+        parameters = (
+            f"client_id={self.client_id}"
+            f"&response_type=code"
+            f"&scope={'%20'.join(self.scopes)}"
+            f"&acr_values={' '.join(acr_values)}"
+            f"&nonce={nonce}"
+            f"&state={state}"
+            f"&redirect_uri={self.post_login_callback_url}"
+        )
+        url = f"{self.fc_base_url}{self.fc_authorization_url}?{parameters}"
+        return url, nonce, state
+
+    def get_configuration(self) -> dict:
+        """Return the OpenID configuration."""
+        response = requests.get(f"{self.fc_base_url}/api/v2/.well-known/openid-configuration", timeout=10)
+        if response.status_code != 200:
+            logger.error(f"FranceConnect - Failed to retrieve configuration: {response.status_code}")
+            return {}
+        return response.json()
+
+    def get_access_token(self, code: str) -> dict:
+        """Exchange the authorization code for an access token."""
+        token_url = f"{self.fc_base_url}{self.fc_token_url}"
+        data = {
+            "grant_type": "authorization_code",
+            "code": code,
+            "redirect_uri": self.post_login_callback_url,
+            "client_id": self.client_id,
+            "client_secret": self.client_secret,
+        }
+
+        response = requests.post(token_url, data=data, headers={"Accept": "application/json"}, timeout=10)
+
+        if response.status_code != 200:
+            logger.error(f"FranceConnect - Failed to retrieve access_token: {response.status_code}")
+            return {}
```

Comment:
You should raise a different custom exception at each step so that codes using your package can handle each failure as they want.

It's also good practice to have a parent exception for the whole package so that all exceptions can be caught by the same `except` if need, something like:

```python
# france_connect/exceptions.py

class FranceConnectError(Exception):
    pass


class FranceConnectHttpError(FranceConnectError):
    
    def __init__(self, status_code: int, body: str):
        self.status_code = status_code
        self.body = body

    def ___str__(self) -> str:
        return f"Request to FranceConnect failed with status code '{status_code}': {body}"


class FranceConnectConfigurationError(FranceConnectHttpError):
     pass


class FranceConnectAccessTokenError(FranceConnectHttpError):
     pass


class FranceConnectUserInfoError(FranceConnectHttpError):
     pass
```